src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -1091,12 +1091,17 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           @inline def tpdPos(transformed: Tree) = typedPos(tree.pos, mode, pt)(transformed)
           @inline def tpd(transformed: Tree)    = typed(transformed, mode, pt)
 
-          @inline def warnValueDiscard(): Unit = if (!isPastTyper && settings.warnValueDiscard) {
+          @inline def warnValueDiscard(): Unit = if (!isPastTyper) {
             def isThisTypeResult = (tree, tree.tpe) match {
               case (Apply(Select(receiver, _), _), SingleType(_, sym)) => sym == receiver.symbol
               case _ => false
             }
```

Comment:
`isThisTypeResult` is meant to detect situations like 

```scala
trait Foo { def meh: this.type }

def test(foo: Foo): Unit = foo.meh
```

and not warn. The assumption is that `this.type` doesn't give you any new information you didn't already have, so it's okay to discard. JSYK.